tgui/packages/tgui/interfaces/MCDependencyDebug.tsx

```diff
@@ -0,0 +1,303 @@
+          >
+            <svg
+              style={{
+                width: '100%',
+                height: '100%',
+                position: 'absolute',
+              }}
+              viewBox="0, 0, 100, 100"
+            >
+              <circle
+                stroke={'blue'}
+                strokeDasharray={`${100 * Math.PI}`}
+                strokeDashoffset={-100 * Math.PI}
+                className={`color-stroke-blue`}
+                strokeWidth="50px"
+                cx="50"
+                cy="50"
+                r="50"
+                fillOpacity="0"
+                transform="rotate(90, 50, 50)"
+              />
+              <circle cx="50" cy="50" r="50" className={`color-fill-blue`} />
+            </svg>
+            <span ref={inputRef} className="ObjectComponent__PortPos" />
+          </Box>
+        </Stack.Item>
+        <Stack.Item grow={1}>{name}</Stack.Item>
+        <Stack.Item>
+          <Box
+            className={classes(['ObjectComponent__Port'])}
+            textAlign="center"
+          >
+            <svg
+              style={{
+                width: '100%',
+                height: '100%',
+                position: 'absolute',
+              }}
+              viewBox="0, 0, 100, 100"
+            >
+              <circle
+                stroke={'blue'}
+                strokeDasharray={`${100 * Math.PI}`}
+                strokeDashoffset={-100 * Math.PI}
+                className={`color-stroke-blue`}
+                strokeWidth="50px"
+                cx="50"
+                cy="50"
+                r="50"
+                fillOpacity="0"
+                transform="rotate(90, 50, 50)"
+              />
+              <circle cx="50" cy="50" r="50" className={`color-fill-blue`} />
+            </svg>
+            <span ref={outputRef} className="ObjectComponent__PortPos" />
+          </Box>
+        </Stack.Item>
+        <Stack.Item />
+      </Stack>
+    </Box>
+  );
+};
+
+type SubsystemRef = {
+  input?: HTMLElement;
+  output?: HTMLElement;
+};
+
+type ConnectionRef = {
+  [key: string]: SubsystemRef;
+};
+
+type SubsystemCoord = {
+  input: Coord;
+  output: Coord;
+};
+
+type ConnectionData = {
+  [key: string]: SubsystemCoord;
+};
+
+type SubsystemLayer = {
+  [subsystemName: string]: number;
+};
+
+type SubsystemMap = {
+  [subsystemName: string]: Subsystem;
+};
+
+function EvaluateSubsystemLayer(
+  subsystem: Subsystem,
+  depth: number,
+  data: SubsystemLayer,
+) {
+  data[subsystem.name] = Math.max(depth, data[subsystem.name] || 0);
+  for (let i = 0; i < subsystem.dependents.length; i++) {
+    EvaluateSubsystemLayer(subsystem.dependents[i], depth + 1, data);
+  }
+}
+
+export const MCDependencyDebug = (props, context) => {
+  const { data } = useBackend<DependencyData>();
+  const { subsystems } = data;
+  const connectionDom = useRef<ConnectionRef>({});
+  const [connectionData, setConnectionData] = useState<ConnectionData>({});
+
+  useLayoutEffect(() => {
```

Comment:
why uselayouteffect over useeffect?

Reply:
useLayoutEffect is supposed to be used for figuring out positioning since it gets called before the browser is painted.
Check the react docs